### PR TITLE
Shutdown support.

### PR DESCRIPTION
**Pull request by rustyrussell**
Including a significant refactor which makes things generally simpler.

### CODE REVIEW

File: lightningd/peer_control.c [hunk 1]
```diff
@@ -183,6 +181,87 @@ static struct io_plan *send_error(struct io_conn *conn,
 	return peer_write_message(conn, pcs, pcs->peer->error, (void *)io_close_cb);
 }
 
+struct getting_gossip_fd {
```
Comment (cdecker) [inline lightningd/peer_control.c]: I agree, this is rather a roundabout way of doing things, but the alternative of handing the peer through `gossipd` seems worse. Maybe we could desynchronize further and allow for the `gossip_fd` to be passed to `channeld` after initialization? That'd duplicate some paths too, but only in `channeld`, which would need to delay some messages.

File: lightningd/peer_control.c [hunk 2]
```diff
@@ -942,6 +944,75 @@ static int peer_got_funding_locked(struct peer *peer, const u8 *msg)
...
+	if (peer->local_shutdown_idx == -1) {
+		u8 *scriptpubkey;
+
+		peer->local_shutdown_idx = wallet_get_newindex(peer->ld);
+		if (peer->local_shutdown_idx == -1) {
+			peer_internal_error(peer,
+					    "Can't get local shutdown index");
+			return -1;
+		}
+		/* FIXME: Save to db */
```
Comment (cdecker) [inline lightningd/peer_control.c]: Not sure what we're supposed to save to the DB here, the index is already incremented and we'll recognize outputs based on that.

File: lightningd/peer_control.c [hunk 3]
```diff
@@ -804,12 +823,10 @@ static enum watch_result funding_lockin_cb(struct peer *peer,
...
 /* FIXME: Reshuffle. */
-static void peer_start_channeld(struct peer *peer, enum peer_state oldstate,
-				const u8 *funding_signed);
+static void peer_start_channeld(struct peer *peer, const u8 *funding_signed);
 
-static bool opening_got_hsm_funding_sig(struct subd *hsm, const u8 *resp,
-					const int *fds,
-					struct funding_channel *fc)
+static void opening_got_hsm_funding_sig(struct funding_channel *fc,
+					const u8 *resp)
```
Comment (rustyrussell) [inline lightningd/peer_control.c]: Sometimes this looks weird in the patch due to the tabs, as in this case.

File: lightningd/peer_control.c [hunk 4]
```diff
@@ -1026,12 +992,13 @@ static int channel_msg(struct subd *sd, const u8 *msg, const int *unused)
...
-/* We've got fd from HSM for channeld */
-static bool peer_start_channeld_hsmfd(struct subd *hsm, const u8 *resp,
-				      const int *fds,
-				      struct peer *peer)
+static bool peer_start_channeld(struct peer *peer,
+				enum peer_state old_state,
+				const u8 *funding_signed)
 {
-	u8 *initmsg;
+	const tal_t *tmpctx = tal_tmpctx(peer);
```
Comment (rustyrussell) [inline lightningd/peer_control.c]: Good catch.  I have a patch in my head to catch such things, too :)

File: lightningd/peer_control.c [hunk 5]
```diff
@@ -1227,9 +1242,17 @@ static bool opening_funder_finished(struct subd *opening, const u8 *resp,
 	tal_free(utxos);
 
 	fc->peer->owner = NULL;
-	peer_set_condition(fc->peer, OPENINGD, GETTING_SIG_FROM_HSM);
-	subd_req(fc, fc->peer->ld->hsm, take(msg), -1, 0,
-		 opening_got_hsm_funding_sig, fc);
+
+	/* FIXME: don't use hsm->conn */
```
Comment (rustyrussell) [inline lightningd/peer_control.c]: We could, but I roll it into one function eventually and it's not too bad.

File: lightningd/peer_control.c (hunk 2 above)
Comment (rustyrussell) [inline lightningd/peer_control.c]: We need to save that it's for this peer: if peer restarts, we need to tell it the shutdown scriptpubkey.

**Comment by cdecker** [comment]
CI tests now pass, they were just flaky...

File: lightningd/peer_control.c [hunk 6]
```diff
@@ -1391,31 +1391,26 @@ static bool gossip_peer_released(struct subd *gossip,
...
-	if (!fromwire_gossipctl_release_peer_reply(resp, NULL, &id, &cs))
-		fatal("Gossup daemon gave invalid reply %s",
-		      tal_hex(gossip, resp));
-
-	/* This is how gossipd handles a reconnect (gossipctl_fail_peer) racing
-	 * with us trying to connect. */
-	if (id != fc->peer->unique_id) {
+	if (!fromwire_gossipctl_release_peer_reply(resp, NULL, &cs)) {
+		if (!fromwire_gossipctl_release_peer_replyfail(resp, NULL)) {
+			fatal("Gossup daemon gave invalid reply %s",
```
Comment (rustyrussell) [inline lightningd/peer_control.c]: Ah, actually that was moving an existing typo!  Which was also mine...

**Pull request merged**